### DISCUSSION TITLE
Reading and writing values for the emergency power supply

### DISCUSSION THREAD

**Discussion by hmichels** [spali/go-rscp#45] [Q&A]
First a big thanks to Spali, I use the integration along with other RSCP integrations but run all services to set values with this integration (command_line). It works easily and very reliable.

Now I want to read values for the emergency power supply and change them again via the "command_line" if necessary. 
In other RSCP integrations for Home Assistant the following tags will be used for this i.e.: EMS_REQ_EP_RESERVE, SE_REQ_SET_EP_RESERVE, SE_PARAM_EP_RESERVE, SE_PARAM_EP_RESERVE_W

If I understand it correctly, this integration uses or builds on the RSCP library "python-e3dc" from "fsantini", right?

Unfortunately, I cannot read or write the above values.
Can anyone help me here?

Thanks a lot!

Greeting,
Holger

**Comment by spali** [comment]
> If I understand it correctly, this integration uses or builds on the RSCP library "python-e3dc" from "fsantini", right?

No, this is a complete independent of any other rscp library. I built it completely from scratch in go. It may include some information/ideas from other libraries as the official documentation lacks a lot.

> Now I want to read values for the emergency power supply and change them again via the "command_line" if necessary.
In other RSCP integrations for Home Assistant the following tags will be used for this i.e.: EMS_REQ_EP_RESERVE, SE_REQ_SET_EP_RESERVE, SE_PARAM_EP_RESERVE, SE_PARAM_EP_RESERVE_W

I don't have this feature on my E3DC, so I can't really test/help here, but it should be possible with this library. You should check the documentation of our model and try to understand the tags and how to query/set it. Maybe another user can help which uses this feature.

**Comment by hmichels** [comment]
Is it possible to use the RSCP library (or the communication with the E3DC system) with "tags" (tag name) or the corresponding number like 0x1B040033 which are not directly implemented in your library (tag.go file) or in the binary?

A nice solution would be to use the developer tools, which can drop the service "shell_command.e3dc_request" with parameter.

I was able to install the binary and the hole communication with the E3DC system inside Home Assistant, by I'm not able to generate a new binary after editing the go files, because I understand neither "go" nor "Linux/Unix".

Can anyone here help who also wants to set the battery reserve for emergency power via RSCP?
Thanks a lot!

**Comment by spali** [comment]
> Is it possible to use the RSCP library (or the communication with the E3DC system) with "tags" (tag name) or the corresponding number like 0x1B040033 which are not directly implemented in your library (tag.go file) or in the binary?

To be honest, currently not sure if I implemented this feature. At least it was planned at first. But try it.
If it works you definilty have to define the complete tuple with datatype of the tag like:
```
[0x00000000, "UChar8", 1]
```

**Comment by hmichels** [comment]
Using the tag number does not work. 
The following service was performed successfully:
```
service: shell_command.e3dc_request
data_template: 
  request: >-
    {{ [['EMS_REQ_SET_POWER_SETTINGS', [
        ['EMS_POWERSAVE_ENABLED', "UChar8", 1]
    ]]] }}
```
with following positive message:
`Thu Oct  5 11:13:44 CEST 2023: [["EMS_REQ_SET_POWER_SETTINGS", [["EMS_POWERSAVE_ENABLED", "UChar8", 1]]]]{"EMS_SET_POWER_SETTINGS":{"EMS_RES_POWERSAVE_ENABLED":0}}`


Now the tag `EMS_REQ_SET_POWER_SETTINGS` was replaced with `0x0100008C` or `'0x0100008C'` and resulted in the following error message in each case:
```
Thu Oct  5 11:56:03 CEST 2023: [[16777356, [["EMS_POWERSAVE_ENABLED", "UChar8", 1]]]]error: Tag should be a string, got 16777356
Thu Oct  5 11:57:17 CEST 2023: [["0x0100008C", [["EMS_POWERSAVE_ENABLED", "UChar8", 1]]]]error: 0x0100008C does not belong to Tag values
```

Therefore, I think that the use of unimplemented tag numbers are not supported in the rscp library 
and therefore an attempt to set the emergency power reserves in this way is not possible, unfortunately.

**Reply by spali** [reply]
True... couldn't remember if I implemented it (was long time ago).
Was/Is on the TODO list especially as there seems different tags with same value for different models. Was long time ago checked to code of the library. But couldn't find someone with different models/tags which could contribute to the library 😞
